Remove deprecated use of `PySys_SetArgvEx`

## Is your feature request related to a problem? Please describe.

Following the update of Python version from 3.9.10 to 3.12.10, building Slicer reports warnings like the following:

```
/path/to/Slicer/Base/QTCore/qSlicerCoreApplication.cxx: In member function ‘virtual void qSlicerCoreApplication::handleCommandLineArguments()’:
/path/to/Slicer/Base/QTCore/qSlicerCoreApplication.cxx:1240:65: warning: ‘void PySys_SetArgvEx(int, wchar_t**, int)’ is deprecated [-Wdeprecated-declarations]
 1240 |     PySys_SetArgvEx(pythonArgc, pythonArgv, /*updatepath=*/false);
      |                                                                 ^
In file included from /path/to/Slicer-Release/python-install/include/python3.12/Python.h:96,
                 from /path/to/Slicer-Release/CTK-build/CMakeExternals/Install/include/PythonQt/PythonQtPythonInclude.h:102,
                 from /path/to/Slicer-Release/CTK-build/CMakeExternals/Install/include/PythonQt/PythonQt.h:45,
                 from /path/to/Slicer/Base/QTCore/qSlicerCoreApplication.cxx:67:
/path/to/Slicer-Release/python-install/include/python3.12/sysmodule.h:14:38: note: declared here
   14 | Py_DEPRECATED(3.11) PyAPI_FUNC(void) PySys_SetArgvEx(int, wchar_t **, int);
      |                                      ^~~~~~~~~~~~~~~
```

## Describe the solution you'd like

Update `qSlicerCoreApplication.cxx` to replace use of `PySys_SetArgvEx` with `PyConfig.argv` as suggested in https://docs.python.org/3/c-api/init.html#c.PySys_SetArgvEx

## Describe alternatives you've considered

NA

## Additional context

* https://github.com/Slicer/Slicer/pull/8456
* https://docs.python.org/3/c-api/init_config.html#init-config

--- a/Base/QTCore/qSlicerCoreApplication.cxx
+++ b/Base/QTCore/qSlicerCoreApplication.cxx
@@ -20,6 +20,10 @@
 
 // standard library includes
 #include <clocale>
+#include <stdexcept>
+
+// Python include header
+#include <Python.h>
 
 // Qt includes
 #include <QDebug>
@@ -669,8 +673,8 @@ QSettings* qSlicerCoreApplicationPrivate::instantiateSettings(bool useTmp)
 #ifdef Slicer_STORE_SETTINGS_IN_APPLICATION_HOME_DIR
   // If a Slicer.ini file is available in the home directory then use that,
   // otherwise use the default one in the user profile folder.
-  // Qt appends organizationName/organizationDomain to the directory set in QSettings::setPath, therefore we must include it in the folder name
-  // (otherwise QSettings() would return a different setting than app->userSettings()).
+  // Qt appends organizationName/organizationDomain to the directory set in QSettings::setPath, therefore we must
+  // include it in the folder name (otherwise QSettings() would return a different setting than app->userSettings()).
   QString iniFileName = QDir(this->SlicerHome).filePath(QString("%1/%2.ini").arg(ctkAppLauncherSettings().organizationDir()).arg(q->applicationName()));
   if (QFile(iniFileName).exists())
   {
@@ -1210,8 +1214,47 @@ void qSlicerCoreApplication::handleCommandLineArguments()
       pythonArgv[i + 1] = QStringToPythonWCharPointer(scriptArgs.at(i));
     }
 
-    // See https://docs.python.org/c-api/init.html
-    PySys_SetArgvEx(pythonArgc, pythonArgv, /*updatepath=*/false);
+    // https://docs.python.org/3/c-api/init_config.html#init-config
+
+    int status_exit_code = 0;
+
+    PyConfig config;
+    PyConfig_InitPythonConfig(&config);
+    config.isolated = 1;
+
+    /* Decode command line arguments.
+       Implicitly preinitialize Python (in isolated mode). */
+    PyStatus status = PyConfig_SetArgv(&config, pythonArgc, pythonArgv);
+    if (PyStatus_Exception(status))
+    {
+      PyConfig_Clear(&config);
+      if (PyStatus_IsExit(status))
+      {
+        status_exit_code = status.exitcode;
+      }
+      /* Display the error message and exit the process with
+         non-zero exit code */
+      Py_ExitStatusException(status);
+    }
+
+    status = Py_InitializeFromConfig(&config);
+    if (PyStatus_Exception(status))
+    {
+      PyConfig_Clear(&config);
+      if (PyStatus_IsExit(status))
+      {
+        status_exit_code = status.exitcode;
+      }
+      /* Display the error message and exit the process with
+         non-zero exit code */
+      Py_ExitStatusException(status);
+    }
+    PyConfig_Clear(&config);
+
+    if (status_exit_code != 0)
+    {
+      throw std::runtime_error("Python initialization failed with status_exit_code" + std::to_string(status_exit_code));
+    }
 
     // Set 'sys.executable' so that Slicer can be used as a "regular" python interpreter
     this->corePythonManager()->executeString(

--- a/Base/QTCore/qSlicerCorePythonManager.cxx
+++ b/Base/QTCore/qSlicerCorePythonManager.cxx
@@ -47,7 +47,17 @@ qSlicerCorePythonManager::qSlicerCorePythonManager(QObject* _parent)
 
   // If it applies, disable import of user site packages
   QString noUserSite = qgetenv("PYTHONNOUSERSITE");
-  Py_NoUserSiteDirectory = noUserSite.toInt();
+
+  PyConfig config;
+  PyConfig_InitPythonConfig(&config);
+
+  config.user_site_directory = noUserSite.toInt(); // disable user site packages
+
+  PyStatus status = Py_InitializeFromConfig(&config);
+  if (PyStatus_Exception(status))
+  {
+    Py_ExitStatusException(status);
+  }
 
   // Import site module to ensure the 'site-packages' directory
   // is added to the python path. (see site.addsitepackages function).

